File: cpp/1-arrays-and-strings/1-1.cpp
#include <iostream>
#include <set>
#include <string>
#include <assert.h> 

using namespace std;

bool allUnique(string s) {
    string sSeen;
    for (auto const& e : s) { 
        if (sSeen.find(e) != std::string::npos) {
            return false;
        }
        else {
            sSeen += e;
        }
    }
    return true;
}

int main() {
    assert(allUnique("") == true);
    assert(allUnique(" ") == true);
    assert(allUnique("Hello, world!") == false);
    assert(allUnique("  ") == false);
    assert(allUnique("Python") == true);
    return 0;
}

File: python/1-arrays-and-strings/1-1.py
# Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

def all_unique(s: str) -> bool:
    if len(s) <= 1:
        return True
    temp = set(s)
    return len(s) == len(temp)


assert all_unique("") == True
assert all_unique(" ") == True
assert all_unique("Hello, world!") == False
assert all_unique("  ") == False
assert all_unique("Python") == True


def all_unique_2(s: str) -> bool:
    for i, c in enumerate(s):
        if c in s[i + 1 :]:
            return False
    return True


assert all_unique_2("") == True
assert all_unique_2(" ") == True
assert all_unique_2("Hello, world!") == False
assert all_unique_2("  ") == False
assert all_unique_2("Python") == True
